app/main/services/search_service.py

```diff
@@ -133,14 +140,39 @@ def url_for_search(**kwargs):
         }
 
         if aggregations:
-            response['aggregations'] = {}
             # Return aggregations in a slightly cleaner format.
-            for k, v in res.get('aggregations', {}).items():
-                response['aggregations'][k] = {d['key']: d['doc_count'] for d in v['buckets']}
+            response['aggregations'] = {
+                k: {d['key']: d['doc_count'] for d in v['buckets']}
+                for k, v in res.get('aggregations', {}).items()
+            }
+
+        # determine whether we're actually off the end of the results. ES handles this as a result-less-yet-happy
+        # response, but we probably want to turn it into a 404 not least so we can match our behaviour when fetching
+        # beyond the `max_result_window` below
+        if search and constructed_query.get("from") and not response["documents"]:
+            return _page_404_response(query_args.get("page", None))
 
         return response, 200
 
     except TransportError as e:
+        root_causes = getattr(e, "info", {}).get("error", {}).get("root_cause", {})
+        if root_causes and root_causes[0].get("reason").startswith("Result window is too large"):
+            # in this case we have to fire off another request to determine how we should handle this error...
+            # (note minor race condition possible if index is modified between the original call and this one)
+            try:
+                result_count = es.count(
+                    index=index_name,
+                    doc_type=doc_type,
+                    body=construct_query(mapping, query_args, page_size=None),
+                )["count"]
+            except TransportError as e:
+                return _get_an_error_message(e), e.status_code
+            else:
```

Comment:
As the `except` contains `return`, I don't think the `else` clause here really helps readability? I suppose this is preference rather than anything concrete.

Reply:
Actually the subtle reason I did this I think was that it pulled the line `return _get_an_error_message(e), e.status_code` up away from its cousin `return _get_an_error_message(e), e.status_code` below it. Seeing the line repeated twice superficially looks like a mistake and might tempt someone to try merging the two lines, which would of course do the wrong thing.

Or maybe it was to do with lowering the function's mccabe measure... can't remember.

Reply:
Oh I see what you mean - could unindent the `else` entirely.... mmmmm still prefer the explicitness. Flow scans better.